### ISSUE TITLE
Add nerdctl (instead of crictl)

### ISSUE THREAD

**Issue by neo-liang-sap** [gardener/ops-toolbelt#56]
per discussion in https://github.com/gardener/ops-toolbelt/pull/21#issuecomment-871366880 , will add nerdctl 
/assign

**Comment by vlerenc** [comment]
/reopen Can we please improve the default experience with `nerdctl`:

When I launch it from a non-`chroot` pod (just to show that it's there - this way it's useless anyway):
```
$ nerdctl ps
FATA[0000] cannot access containerd socket "/run/containerd/containerd.sock": no such file or directory
```

When I launch it from a `chroot` pod:
```
$ nerdctl
bash: nerdctl: command not found
```

How are you doing that? https://github.com/gardener/ops-toolbelt/blob/master/hacks/ops-pod isn't doing any magic here, I believe. Who is using this actively and how, please?

cc @neo-liang-sap @plkokanov @jfortin-sap
cc @dguendisch @hendrikKahl @BeckerMax

**Comment by plkokanov** [comment]
You can check with `ghelp` from the container. There should be an entry for `nerdctl` that says:
`nerdctl is a Docker-compatible CLI for containerd. If the root directory of the host is mounted you can use the files under /hostroot/run/ to connect to the container runtime of the host e.g. 'nerdctl --address unix:///hostroot/run/containerd/contaierd.sock ps'`  :blush:  . But I haven't done that in a while so not sure if it will work as expected.

Using `chroot` basically makes the root file system of the node to also be the root fs of the container so there will be no `nerdctl` available on the `PATH` or accessible in any way. We could potentially copy over some clis to the node's file system, but I'm not sure if that would be ok.

**Comment by vlerenc** [comment]
@plkokanov Are these just some comments in the direction of whoever will help/improve, because otherwise I don't understand the intent.

I required a CLI to do ops and had none resp. the experience was poor. It doesn't work with the current images and hack/launch scripts (works as shown above). How are people doing their ops work then? Everybody is helping themselves instead of the toolbelt to provide a good out-of-the-box ops experience? In previous times, I had the `docker` CLI, now we recognised that we need to do something with the transition away from Docker, but it was done only half-way and that's what this ticket is about in the hope to improve our work experience.

**Comment by plkokanov** [comment]
The first part of my comment was about how you can use `nerdctl` currently. If you use the `ops-pod` command, without specifying the `--chroot` flag, then the root directory of the node on which the `ops-pod` is scheduled is mounted under `/host`. So if the host uses `containerd` you can run the following command to list all containers:
```
$ nerdctl --address /host/run/docker/containerd/containerd.sock --namespace moby ps
CONTAINER ID    IMAGE    COMMAND                   CREATED           STATUS    PORTS    NAMES
01c5f5a7eb15             "/aws-cloud-controll…"    31 minutes ago    Up                     
04b61763daba             "/pause"                  31 minutes ago    Up                     
079775b7f41e             "/usr/local/bin/pilo…"    2 weeks ago       Up                     
0ad8a4c610ef             "/run.sh"                 13 days ago       Up    
```


I didn't mean that we shouldn't improve the situation, so the second part of my comment was a suggestion on how to do that.  So here are a few additional ideas that could work out better:
1. Make it so that the host's root directory is always mounted under the same name for pods started by the gardener dashboard terminal functionality and the `ops-pod` command. Currently the former uses `/hostroot` the latter `/host`
2. Either try to detect or use some default setting for `nerdctl`'s `--address` flag and use a wrapper script or an alias so that `--address` does not have to be specified manually every time.
3. If 2. doesn't work out, improve the documentation for `nerdctl` that is presented by the `ghelp` command so that it is clearer how one can find the `.sock` file to use for `nerdctl`'s `--address` flag

**Comment by vlerenc** [comment]
A possible 4. solution: Have an alias.

However, when I inspect the containers from underneath, I usually do that with all `host` permissions/`chroot`'ed, so that it feels more like being on the node directly.

**Comment by vlerenc** [comment]
Today I was again in the situation to use `crictl`/`nerdctl`. While it remains somewhat inconvenient, I understood that no changes are planned, so can we close this ticket then?

BTW @neo-liang-sap @petersutter I believe `nerdctl` is outdated? I got no volume information and read that it's now supported also in the default/docker mode since April/May: https://github.com/containerd/nerdctl/issues/483.

**Comment by petersutter** [comment]
@vlerenc which version of the ops-toolbelt are you using? I ran the ops toolbelt from the dashboard webterminals and using the `/hacks/ops-pod` script. Running `nerdctl ps` works without issues (no chroot).

regarding https://github.com/containerd/nerdctl/issues/483. This was fixed with https://github.com/containerd/nerdctl/pull/822, which was part of nerdctl [`v0.19.0`](https://github.com/containerd/nerdctl/releases/tag/v0.19.0). The latest ops-toolbelt release includes nerdctl `v0.22.0`. So this feature should be part of it. 
Of course we can also bump it to the latest available version `v0.23.0`.

**Comment by vlerenc** [comment]
> @vlerenc which version of the ops-toolbelt are you using? I ran the ops toolbelt from the dashboard webterminals and using the /hacks/ops-pod script. Running nerdctl ps works without issues (no chroot).

Yes, it works in a regular pod. But when I really need it, I am in a node-level chroot'ed pod.

Anyway, I have my own launch scripts since ages anyhow and changed it / have now an alias (as I suggested above) to the binary, so that it works for me, but I doubt it does for operators in general when [looking here](https://github.com/gardener/ops-toolbelt/blob/master/hacks/ops-pod).

**Comment by vlerenc** [comment]
> regarding https://github.com/containerd/nerdctl/issues/483. This was fixed with https://github.com/containerd/nerdctl/pull/822, which was part of nerdctl [v0.19.0](https://github.com/containerd/nerdctl/releases/tag/v0.19.0). The latest ops-toolbelt release includes nerdctl v0.22.0. So this feature should be part of it. Of course we can also bump it to the latest available version v0.23.0.

Well, maybe I was wrong. I believe to use latest, something like `eu.gcr.io/gardener-project/gardener/ops-toolbelt`. Wrong one? And there I described the containers and volumes were missing, so it used the raw mode and saw then.

Anyway, not worth following up (for me). I am good, but was wondering how everybody else works with this.

Shall we close this ticket now?

**Comment by petersutter** [comment]
yes `eu.gcr.io/gardener-project/gardener/ops-toolbelt` should get you the latest image, strange.

> Anyway, not worth following up (for me). I am good, but was wondering how everybody else works with this.
> 
> Shall we close this ticket now?

At least the title of the issue is misleading for the current issue.
Can't tell if it's something that operators are missing / have troubles with. The others that were mentioned in this issue haven't commented. Seems not to be an issue for them? So let's /close it..
